Template Authorization API

varshavaradarajan opened the pull request.

maheshp commented on server/src/main/java/com/thoughtworks/go/config/update/UpdateTemplateAuthConfigCommand.java:
```diff
@@ -0,0 +1,40 @@
+        PipelineTemplateConfig existingTemplateConfig = findAddedTemplate(modifiedConfig);
+        TemplatesConfig templatesConfig = modifiedConfig.getTemplates();
+        existingTemplateConfig.setAuthorization(templateConfig.getAuthorization());
+        modifiedConfig.setTemplates(templatesConfig);
```
I couldn't understand the need to get templates config and set it back again, is this required?

varshavaradarajan replied:
No, templates need not be set. I'll change it.

maheshp commented on server/webapp/WEB-INF/rails.new/app/controllers/api_v1/admin/authorization/templates_controller.rb:
```diff
@@ -0,0 +1,67 @@
+          handle_create_or_update_response(result, updated_template)
+        end
+
+        def load_template(template_name = params[:template_name])
```
Move these methods to private?

maheshp commented on server/webapp/WEB-INF/rails.new/app/controllers/api_v1/admin/authorization/templates_controller.rb:
```diff
@@ -0,0 +1,67 @@
+        def update
+          result = HttpLocalizedOperationResult.new
+          authorization = ApiV1::Admin::Authorization::AuthorizationConfigRepresenter.new(com.thoughtworks.go.config.Authorization.new).from_hash(params[:template])
+          updated_template = CLONER.deepClone(@template)
```
How bout getting rid of the template cloning? 
change `template_config_service.updateTemplateAuthConfig(current_user, template_name, auth_config, result, etag)'
and UpdateTemplateAuthConfigCommand update can update the auth_config as part of the update method.

What do you think?

maheshp commented:
Listing out few things discussed with @varshavaradarajan 

1. Since authorization is a nested resource of templates we are considering changing the routes to `/go/api/admin/templates/<template_id>/authorization`.
2. Change templates_controller to authorization_controller, namespace within templates.
3. templates authorization API will expose only GET and PUT. The idea is when a template is created it would have an empty authorization, so creating authorization is not required but can be updated.

maheshp commented on server/webapp/WEB-INF/rails.new/config/routes.rb:
```diff
@@ -241,6 +241,12 @@
         end
+
+        namespace :templates do
+          get ':template_name/authorization' => 'authorization#show', constraints: {template_name: ALLOW_DOTS}
```
Use `TEMPLATE_NAME_FORMAT` for template_name contraints.

maheshp commented on server/webapp/WEB-INF/rails.new/app/presenters/api_v1/admin/authorization/authorization_config_representer.rb:
```diff
@@ -0,0 +1,46 @@
+                 decorator: ApiV1::Admin::Authorization::UserRoleConfigRepresenter,
+                 class: com.thoughtworks.go.config.AdminsConfig
+
+        property :operationConfig, as: :operate,
```
Should we expose this since templates authorisation do not support `operate`?

varshavaradarajan commented:
PR build went through -  https://build.gocd.org/go/pipelines/value_stream_map/trigger/2238 (regression SPA was a flaky one). Just rebased it now. 

@maheshp - we discussed about why the authorization validate just returns without performi. It because it will end up doing the same validation twice when using the GoConfigGraphWalker. Its the same reason why I removed `validateStageConfig` from `PipelineTemplateConfig#validate`.

varshavaradarajan commented:
Reason why DelegatingValidationContext was introduced

Consider validation of the roles configured in template authorization and template stage config approval with ConfigSaveValidationContext. 

Roles configured in template authorization should be validated to check if they exist. 
Roles configured in **template** stage config should **not** be validated as this could be a parameter.

Given the above 2 restrictions, using just the `ConfigSaveValidationContext#isWithTemplates()` is not enough. This method would have returned true for template authorization as well and skipped validation of template authorization. Therefore, we introduced the DelegatingValidationContext. 

Here's the PR that  introduced it - https://github.com/gocd/gocd/pull/3035
